File: Stack/232. Implement Queue using Stacks /ImplementQueueUsingStacks.cpp
//
//  ImplementQueueUsingStacks.cpp
//  LeetCode-main
//
//  Created by jackma on 2022/2/27.
//  Copyright © 2022 JackMa. All rights reserved.
//

#include "ImplementQueueUsingStacks.hpp"

/**
 Implement a first in first out (FIFO) queue using only two stacks. The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).
 
 Input
 ["MyQueue", "push", "push", "peek", "pop", "empty"]
 [[], [1], [2], [], [], []]
 Output
 [null, null, null, 1, 1, false]
 */

// my solution
MyQueue::MyQueue() {
    myStack = stack<int>();
}

void MyQueue::push(int x) {
    stack<int> tempStack;
    while (myStack.size() != 0) {
        tempStack.push(myStack.top());
        myStack.pop();
    }
    
    myStack.push(x);
    while (tempStack.size() != 0) {
        myStack.push(tempStack.top());
        tempStack.pop();
    }
}

int MyQueue::pop() {
    int pop = myStack.top();
    myStack.pop();
    return pop;
}

int MyQueue::peek() {
    return myStack.top();
}

bool MyQueue::empty() {
    return myStack.empty();
}

void testMyQueue() {
    MyQueue queue = MyQueue();
    queue.push(1);
    queue.push(2);
    cout << queue.peek() << " ";
    cout << queue.pop() << " ";
    cout << queue.empty() << endl;
}

File: Stack/232. Implement Queue using Stacks /ImplementQueueUsingStacks .swift
//
//  ImplementQueueUsingStacks .swift
//  LeetCode-Swift
//
//  Created by jackma on 2023/3/12.
//  Copyright © 2023 JackMa. All rights reserved.
//

import Foundation

/**
 https://leetcode.com/problems/implement-queue-using-stacks
 
 Implement a first in first out (FIFO) queue using only two stacks. The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).
 */

class MyQueue {
    var stack: Array<Int>
    
    init() {
        stack = [Int]()
    }
    
    func push(_ x: Int) {
        stack.insert(x, at: 0)
    }
    
    func pop() -> Int {
        guard let pop = stack.last else { return 0 }
        stack.removeLast()
        return pop
    }
    
    func peek() -> Int {
        guard let pop = stack.last else { return 0 }
        return pop
    }
    
    func empty() -> Bool {
        return stack.count == 0
    }
}
